### PR TITLE
Implement a crm_mon XML reader

### PR DESCRIPTION
**Pull request by jgrund**
Implement a reader for `crm_mon -1 -r -X`.

It uses streaming XML parsing to build up a `Cluster` struct.
The `Cluster` struct just contains a list of `Nodes` for now, but can be
expanded in the future.

In addition, use thiserror for iml_agent errors.

Signed-off-by: Joe Grund <jgrund@whamcloud.io>

---
This change is [Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2182)

### CODE REVIEW

File: iml-agent/src/crm_reader.rs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+        "true" => Ok(true),
+        "false" => Ok(false),
+        _ => Err(ImlAgentError::Io(io::Error::new(
+            io::ErrorKind::NotFound,
+            "Could not convert to bool",
+        ))),
+    }
+}
+
+fn from_map(x: &HashMap<&str, String>) -> Result<Node, ImlAgentError> {
```
Comment (utopiabound) [inline iml-agent/src/crm_reader.rs:55]: should this be `node_from_map`?

Comment (jgrund) [reply iml-agent/src/crm_reader.rs:55]: sure, I'll update.

**Comment by jgrund** [comment]
Ref: #2139

File: iml-agent/src/crm_reader.rs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    x.get(arg)
+        .map(|x| x.as_str())
+        .ok_or_else(|| ImlAgentError::MissingArgument(arg.into()))
+}
+
+fn required_bool(arg: &str, x: &HashMap<&str, String>) -> Result<bool, ImlAgentError> {
+    required_arg(arg, x).and_then(str_to_bool)
+}
+
+fn str_to_bool(x: &str) -> Result<bool, ImlAgentError> {
```
Comment (johnsonw) [inline iml-agent/src/crm_reader.rs:44]: you could use std::str::FromStr:
```
let b: Result<bool, ParseBoolError> = std::str::FromStr::from_str("true");
```
https://doc.rust-lang.org/std/primitive.bool.html#impl-FromStr

**Review by johnsonw — DISMISSED** [review]

File: iml-agent/src/crm_reader.rs (hunk 2 above)
Comment (jgrund) [reply iml-agent/src/crm_reader.rs]: Updated to use `.parse`

File: iml-agent/src/crm_reader.rs [hunk 3]
```diff
@@ -38,18 +38,10 @@ fn required_arg<'a>(arg: &str, x: &'a HashMap<&str, String>) -> Result<&'a str,
 }
 
 fn required_bool(arg: &str, x: &HashMap<&str, String>) -> Result<bool, ImlAgentError> {
-    required_arg(arg, x).and_then(str_to_bool)
-}
+    let x = required_arg(arg, x)?;
+    let x = x.parse()?;
```
Comment (johnsonw) [inline iml-agent/src/crm_reader.rs:42]: Could this just be:
```
x.parse()
```
Since it returns a result?

Comment (jgrund) [reply iml-agent/src/crm_reader.rs:42]: No, because the errors wouldn't match. `?` is doing a conversion to `ImlAgentError`.

Comment (johnsonw) [reply iml-agent/src/crm_reader.rs:42]: 👍

**Comment by jgrund** [comment]
Retest in: https://build.whamcloud.com/view/IML/job/manager-for-lustre/1849/

**Review by mkpankov — APPROVED** [review]
> Reviewed 9 of 11 files at r1, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/whamcloud/integrated-manager-for-lustre/2182)* status: all files reviewed, 3 unresolved discussions (waiting on @jgrund, @johnsonw, and @utopiabound)

**Review by ip1981 — APPROVED** [review]

**Pull request merged**